Add runtime stats to sys_mem_blocks

peter-mitsis opened the pull request:
This set of changes adds runtime statistics (and supporting test) to the sys_mem_blocks (lib/os/mem_blocks.c). This feature is enabled when CONFIG_SYS_MEM_BLOCKS_RUNTIME_STATS=y. The interface was modelled after the heap runtime statistics, so each query will return the following information.
  1. Free space
  2. Allocated space
  3. Maximum allocated space.

Memory block statistics can be queried using sys_mem_blocks_runtime_stats_get(). Like the heap runtime statistics, the maximum allocated space can be reset to the currently allocated value (see sys_mem_blocks_runtime_stats_reset_max).

Update:
A few new commits have been added to this set.
A. The previously named 'sys_mem_blocks_runtime_stats' structure has been renamed to a more generic 'sys_memory_stats'.
B. Since the contents of this structure is identical to the heaps 'sys_heap_runtime_stats' structure, the heap version has been removed and replaced with the generic 'sys_memory_stats'.
C. Finally, to round things out and give some consistency to the stats retrieval, the mem_slabs has been updated to add stats_get and reset_max routines that utilize the generic 'sys_memory_stats' structure.

peter-mitsis committed:
lib/os: add statistics tracking to mem_blocks

Both the current and maximum number of allocations in a given memory
blocked are tracked (and can be queried) when the
CONFIG_SYS_MEM_BLOCKS_RUNTIME_STATS Kconfig option is selected.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
tests: Add mem_blocks_stats test

The mem_blocks_stats test verifies the behaviour of the mem_blocks
runtime stats.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

andyross commented on include/zephyr/sys/mem_blocks.h:
```diff
@@ -80,6 +80,14 @@ typedef sys_mem_blocks_t *(*sys_multi_mem_blocks_choice_fn_t)
  */
 
+#ifdef CONFIG_SYS_MEM_BLOCKS_RUNTIME_STATS
+struct sys_mem_blocks_runtime_stats {
```
Can't we just use the heap stats struct directly?  (Ditto for slabs too, FWIW).  The allocator behavior is different, but the information collected seems identical.

peter-mitsis replied:
I was originally thinking that it may be prudent to keep them separate in case the info to track diverges. However, the more I think about it, the more I agree that it would be better to have them report using the same structure. If they do diverge, then unused fields could be set to -1, or some other standard value. After all, it is not like there isn't precedence elsewhere--I'm thinking of "struct stat" in particular.

I'm planning on making this change.

peter-mitsis committed:
lib/os: use generic mem stats structure for heap

Since the retrieved heap memory statistics are identical to those of
the retrieved mem_block statistics, it makes sense to use a single
generic-named structure for both instead of two identical structures
with different names.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Extend slabs memory usage stats

Adds memory usage runtime stats routines that parallel those used
by both the heap and mem_blocks. This helps maintain some level of
of consistency across the different memory types.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
tests: Add mslab_stats test

This test case verifies that the mem slab runtime stats routines
behave as expected.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

carlescufi commented:
@andyross could you please take another look?

andyross commented on tests/kernel/mem_slab/mslab_stats/src/main.c:
```diff
@@ -0,0 +1,170 @@
+			 ztest_unit_test(test_mem_slab_runtime_stats)
+			 );
+
+	ztest_run_test_suite(lib_mem_slab_stats_test);
```
This is crossing with the giant "ZTEST_NEW_API" switchover.  Not a problem for me personally, but I'm sure someone's going to sneak in with a -1 and ask you do use the new stuff.

andyross reviewed (dismissed):
Still looks good.  Quibble about the test API, so might want to merge it fast and fix it later. :)

peter-mitsis commented on tests/kernel/mem_slab/mslab_stats/prj.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_ZTEST=y
```
Updated.

andyross reviewed (approved):
Still +1

carlescufi commented:
@yperess please take another look

yperess commented on tests/kernel/mem_slab/mslab_stats/src/main.c:
```diff
@@ -0,0 +1,162 @@
+		      status, -EINVAL);
+}
+
+ZTEST(lib_mem_slab_stats_test, test_mem_slab_runtime_stats)
```
More of a nit (I'm not going to block this change for this) but this is testing too many things. It should really be multiple tests.
